### PR TITLE
[Observables] chore: add `ReplayObservable#SubscribeFromLatestBufferedOffset()`

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Add `ReplayObservable#SubscribeFromLatestBufferedOffset()` to support finer control over replay observable subscriptions.

## Issue

- #553 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [x] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to subscribe from a specified offset in the replay buffer, improving flexibility in data observation.
- **Tests**
	- Added tests to verify the new subscription functionality from the latest buffered offset.

### CODE REVIEW

File: pkg/observable/interface.go [hunk 1]
```diff
@@ -11,6 +11,14 @@ import "context"
 // to new observers, before publishing new values to observers.
 type ReplayObservable[V any] interface {
 	Observable[V]
+	// SubscribeFromLatestBufferedOffset returns an observer which is initially notified of
+	// values in the replay buffer, starting from the latest buffered value index - offset.
+	// After this range of the replay buffer is notified, the observer continues to be notified,
+	// in real-time, when the publishCh channel receives a value. If offset is greater than
+	// replayBufferSize or the number of elements it currently contains, the observer is notified
+	// of all elements in the replayBuffer, starting from the beginning. Passing 0 for offset
+	// is equivalent to calling Subscribe() on a non-replay observable.
```
Comment (Olshansk) [inline pkg/observable/interface.go:20]:
```suggestion
	// SubscribeFromLatestBufferedOffset returns an observer which is initially notified of
	// values in the replay buffer, starting from the latest buffered value at index 'offset'.
	//
	// After this range of the replay buffer is notified, the observer continues to be notified,
	// in real-time, when the publishCh channel receives a new value.
	// 
	// If offset is greater than replayBufferSize or the number of elements it currently contains,
	// the observer is notified of all elements in the replayBuffer, starting from the beginning.
	//
	// Passing 0 for offset is equivalent to calling Subscribe() on a non-replay observable.
```

File: pkg/observable/channel/replay_test.go [hunk 2]
```diff
@@ -300,3 +301,78 @@ func TestReplayObservable_Last_Blocks_And_Times_Out(t *testing.T) {
...
+		{
+			name:             "endOffset > replayBufferSize",
+			replayBufferSize: 8,
+			endOffset:        10,
+			expectedValues:   values[2:],
+		},
+		{
+			name:             "replayBufferSize < eldOffset < numBufferedValues ",
+			replayBufferSize: 20,
+			endOffset:        15,
```
Comment (Olshansk) [inline pkg/observable/channel/replay_test.go:336]: 1. WHat's eldOffset?
2. Assuming (1) is a typo of end, the `<` is still incorrect

File: pkg/observable/channel/replay_test.go [hunk 3]
```diff
@@ -300,3 +301,78 @@ func TestReplayObservable_Last_Blocks_And_Times_Out(t *testing.T) {
...
+		name             string
+		replayBufferSize int
+		endOffset        int
+		expectedValues   []int
+	}{
+		{
+			name:             "endOffset = replayBufferSize",
+			replayBufferSize: 8,
+			endOffset:        8,
+			expectedValues:   values[2:], // []int{2, 3, 4, 5, ...},
```
Comment (Olshansk) [inline pkg/observable/channel/replay_test.go:319]:
```suggestion
			expectedValues:   values[2:], // []int{2, 3, 4, 5, ..., 9},
```

File: pkg/observable/channel/replay_test.go [hunk 4]
```diff
@@ -300,3 +301,78 @@ func TestReplayObservable_Last_Blocks_And_Times_Out(t *testing.T) {
...
+			name:             "endOffset < replayBufferSize",
+			replayBufferSize: 10,
+			endOffset:        2,
+			expectedValues:   values[8:], // []int{8, 9},
+		},
+		{
+			name:             "endOffset > replayBufferSize",
+			replayBufferSize: 8,
+			endOffset:        10,
+			expectedValues:   values[2:],
```
Comment (Olshansk) [inline pkg/observable/channel/replay_test.go:331]: I don't understand why this doesn't start at `values[len(values)-1:]`?

File: pkg/observable/channel/replay.go [hunk 5]
```diff
@@ -122,18 +122,40 @@ func (ro *replayObservable[V]) Last(ctx context.Context, n int) []V {
 // It replays the values stored in the replay buffer in the order of their arrival
 // before emitting new values.
 func (ro *replayObservable[V]) Subscribe(ctx context.Context) observable.Observer[V] {
+	return ro.SubscribeFromLatestBufferedOffset(ctx, ro.replayBufferSize)
+}
+
+// SubscribeFromLatestBufferedOffset returns an observer which is initially notified of
```
Comment (Olshansk) [inline pkg/observable/channel/replay.go:128]: Either copy-pasta the updated comment in the interface or reference it.

File: pkg/observable/channel/replay.go [hunk 6]
```diff
@@ -122,18 +122,40 @@ func (ro *replayObservable[V]) Last(ctx context.Context, n int) []V {
...
 	obs, ch := NewObservable[V]()
 	ctx, cancel := context.WithCancel(ctx)
 
 	go func() {
-		// Replay the values stored in the buffer form the oldest to the newest.
 		ro.replayBufferMu.RLock()
-		for i := len(ro.replayBuffer) - 1; i >= 0; i-- {
+
+		// Ensure that the offset is within the bounds of the replay buffer.
+		if endOffset > len(ro.replayBuffer) {
```
Comment (Olshansk) [inline pkg/observable/channel/replay.go:147]: This makes it a bit confusing if endOffset is number of elements from the start or the end of the buffer.

Should we PUC that "it points to the last element in the replayBuffer"?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/observable/channel/replay_test.go [hunk 7]
```diff
@@ -300,3 +301,78 @@ func TestReplayObservable_Last_Blocks_And_Times_Out(t *testing.T) {
...
+		},
+		{
+			name:             "replayBufferSize < eldOffset < numBufferedValues ",
```
Comment (bryanchriswhite) [inline pkg/observable/channel/replay_test.go:334]:
```suggestion
			name:             "replayBufferSize > endOffset > numBufferedValues ",
```

File: pkg/observable/channel/replay_test.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/observable/channel/replay_test.go:331]: It's `values[len(values)-replayBufferSize]`. 

```
// Published values
[ 1 2 3 4 5 ]

// Replay buffer
[ 5 4 3 2 1 ]

// SubscribeFromEndOffset(100)
[ 1 2 3 4 5 ]

// SubscribeFromEndOffset(5)
[ 1 2 3 4 5 ]

// SubscribeFromEndOffset(3)
[ 3 4 5 ]
```

File: pkg/observable/channel/replay.go [hunk 8]
```diff
@@ -122,18 +122,40 @@ func (ro *replayObservable[V]) Last(ctx context.Context, n int) []V {
...
+}
+
+// SubscribeFromLatestBufferedOffset returns an observer which is initially notified of
+// values in the replay buffer, starting from the latest buffered value index - offset.
+// After this range of the replay buffer is notified, the observer continues to be notified,
+// in real-time, when the publishCh channel receives a value. It also returns the current
+// replayBufferSize. If offset is greater than replayBufferSize or the number of elements it
+// currently contains, the observer is notified of all elements in the replayBuffer,
+// starting from the beginning. Passing 0 for offset is equivalent to calling
+// Subscribe() on a non-replay observable.
```
Comment (bryanchriswhite) [inline pkg/observable/channel/replay.go:135]:
```suggestion
// SubscribeFromLatestBufferedOffset returns an observer which is initially notified of
// values in the replay buffer, starting from the latest buffered value index - offset.
// After this range of the replay buffer is notified, the observer continues to be notified,
// in real-time, when the publishCh channel receives a value. If offset is greater than
// replayBufferSize or the number of elements it currently contains, the observer is notified
// of all elements in the replayBuffer, starting from the beginning. Passing 0 for offset
// is equivalent to calling Subscribe() on a non-replay observable.
```

File: pkg/observable/channel/replay.go (hunk 6 above)
Comment (bryanchriswhite) [reply pkg/observable/channel/replay.go:147]: It's an offset from the "latest buffered" value. I thought the name and godoc comment would make it sufficiently clear. I'm not sure I understand your suggestion; what is pointing to the last element in the replayBuffer? And by last do you mean "latest" or largest index?

**Review by red-0ne — APPROVED** [review]
Love it!

**Pull request merged**